packages/openapi3/src/schema-emitter.ts

```diff
@@ -0,0 +1,866 @@
+import {
+  BooleanLiteral,
+  compilerAssert,
+  DiscriminatedUnion,
+  Enum,
+  EnumMember,
+  getDeprecated,
+  getDiscriminatedUnion,
+  getDiscriminator,
+  getDoc,
+  getEncode,
+  getFormat,
+  getKnownValues,
+  getMaxItems,
+  getMaxLength,
+  getMaxValue,
+  getMaxValueExclusive,
+  getMinItems,
+  getMinLength,
+  getMinValue,
+  getMinValueExclusive,
+  getNamespaceFullName,
+  getPattern,
+  getSummary,
+  getTypeName,
+  ignoreDiagnostics,
+  IntrinsicScalarName,
+  IntrinsicType,
+  isArrayModelType,
+  isNeverType,
+  isNullType,
+  isSecret,
+  isTemplateDeclaration,
+  Model,
+  ModelProperty,
+  NumericLiteral,
+  Program,
+  Scalar,
+  StringLiteral,
+  Tuple,
+  Type,
+  TypeNameOptions,
+  Union,
+  UnionVariant,
+} from "@typespec/compiler";
+import {
+  ArrayBuilder,
+  AssetEmitter,
+  Context,
+  Declaration,
+  EmitEntity,
+  EmitterOutput,
+  ObjectBuilder,
+  Placeholder,
+  Scope,
+  SourceFileScope,
+  TypeEmitter,
+} from "@typespec/compiler/emitter-framework";
+import { getVisibilitySuffix, MetadataInfo, Visibility } from "@typespec/http";
+import {
+  checkDuplicateTypeName,
+  getExtensions,
+  getExternalDocs,
+  getOpenAPITypeName,
+  isReadonlyProperty,
+  shouldInline,
+} from "@typespec/openapi";
+import { getOneOf, getRef } from "./decorators.js";
+import { OpenAPI3EmitterOptions, reportDiagnostic } from "./lib.js";
+import { ResolvedOpenAPI3EmitterOptions } from "./openapi.js";
+import { OpenAPI3Discriminator, OpenAPI3Schema, OpenAPI3SchemaProperty } from "./types.js";
+import { VisibilityUsageTracker } from "./visibility-usage.js";
+
+/**
+ * OpenAPI3 schema emitter. Deals with emitting content of `components/schemas` section.
+ */
+export class OpenAPI3SchemaEmitter extends TypeEmitter<
+  Record<string, any>,
+  OpenAPI3EmitterOptions
+> {
+  #metadataInfo: MetadataInfo;
+  #visibilityUsage: VisibilityUsageTracker;
+  #options: ResolvedOpenAPI3EmitterOptions;
+  constructor(
+    emitter: AssetEmitter<Record<string, any>, OpenAPI3EmitterOptions>,
+    metadataInfo: MetadataInfo,
+    visibilityUsage: VisibilityUsageTracker,
+    options: ResolvedOpenAPI3EmitterOptions
+  ) {
+    super(emitter);
+    this.#metadataInfo = metadataInfo;
+    this.#visibilityUsage = visibilityUsage;
+    this.#options = options;
+  }
+
+  // TODO should we need this for all the types?
+  modelDeclarationReferenceContext(model: Model, name: string): Context {
```

Comment:
Need this technically for every type is there a better way or I need to define the referenceConext for all

Reply:
Wouldn't you also need this for modelLiteralReferenceContext?

Reply:
Also is this needed when you have the parameter to emitType?

Reply:
This checks if the type will change with visibility if not then we just reuse the no visibility context so we don’t make another declaration. 

This might happen within the emit type call when we browse for array which inject the item visibility

Reply:
Which is why technically this should be checked for every type